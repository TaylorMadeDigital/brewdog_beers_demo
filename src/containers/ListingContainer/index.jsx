import React, { useEffect } from 'react';
import { useDispatch, useSelector } from "react-redux";
import axios from 'axios';

import s from './index.scss';
import { CardComponent } from '../../components/CardComponent';
import FilterComponent from '../../components/Filter';
import { sortArray, sortArrayByDate, sortArrayByNumber } from '../../utils/array';
import Header from '../../components/Header';
import { CONTENT } from '../../consts/content';

const apiUrl = 'https://api.punkapi.com/v2/';

function ListingContainer() {

  const beers = useSelector(state => state.beers.all);
  const filter = useSelector(state => state.beers.filter);
  const dispatch = useDispatch();

  useEffect(() => {
    async function fetchBeers() {
      const response = await axios.get(`${apiUrl}beers`);
      await dispatch({ type: 'GET_ALL', payload: response.data });
    } 
      fetchBeers();
  }, [dispatch]);

  const getBeerList = () => {
    if (filter) {
      if (filter === 'name') return sortArray(beers, filter);
      if (filter === 'first_brewed') return sortArrayByDate(beers, filter, 'MM/YYYY');
      return sortArrayByNumber(beers, filter);
    }
    return beers;
  }
  const BeerList = () => {
    return getBeerList().map(beer => {
      const item = {
        id: beer.id,
        title: beer.name,
        subtitle: beer.tagline,
        link: `detail/${beer.id}`,
      };
      return (<CardComponent className="cardItem" key={item.id} item={item}>
        <p>{CONTENT.listing.brewed} {beer.first_brewed}</p>
        <span>{CONTENT.listing.abv} {beer.abv || CONTENT.listing.notAvailable}</span>
        <span>{CONTENT.listing.ibu} {beer.ibu || CONTENT.listing.notAvailable}</span>
      </CardComponent>);
    });
  }
  return (
    <div className={s.root}>
        <Header />
        <FilterComponent />
        <div className={s.cardList}>
          <BeerList />
        </div>
    </div>
  );
}

export default ListingContainer;
