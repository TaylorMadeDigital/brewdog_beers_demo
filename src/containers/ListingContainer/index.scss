@import '../../assets/scss/grid';
@import '../../assets/scss/variables';

.root {
    overflow: hidden;
}
.cardList {
    @include gridContainer;
    padding: $space-base;
}

span {
    margin: $space-s;
    display: inline-block;
}
