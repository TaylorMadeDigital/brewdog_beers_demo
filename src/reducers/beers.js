import { BeerActionTypes } from '../actions/index';

const beers = (state = { all: [], current: {}, filter: null}, action) => {
  switch (action.type) {
    case BeerActionTypes.GET_ALL:
      return Object.assign({}, state, {
        all: action.payload
      })
      case BeerActionTypes.GET_ONE:
        return Object.assign({}, state, {
          current: action.payload
        })
    case BeerActionTypes.FILTER_BY:
      return Object.assign({}, state, {
        filter: action.payload
      })
    default:
      return state
  }
}

export default beers;
