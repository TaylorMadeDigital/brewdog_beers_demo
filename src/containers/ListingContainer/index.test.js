import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, cleanup, screen } from "@testing-library/react";
import { createStore } from 'redux'
import { Provider } from 'react-redux';
import ListingContainer from './index';
import beers from './index.fixtures';

describe('<ListingContainer />', () => {

  afterEach(cleanup);

  it('snapshot: renders default state', () => {
    const store = createStore(() => ({ beers: { all: beers, current: beers[0], filter: null }}));
    const { asFragment } = render(<Provider store={store}><Router><ListingContainer /></Router></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('snapshot: renders ordered by abv', async () => {
    const store = createStore(() => ({ beers: { all: beers, current: beers[0], filter: 'abv' }}));
    const { asFragment } = render(<Provider store={store}><Router><ListingContainer /></Router></Provider>);
    expect(asFragment()).toMatchSnapshot();
    const cards = await screen.findAllByTestId('cardHeading')
    expect(cards[0]).toHaveTextContent('Trashy Blonde');
  });
  
});

