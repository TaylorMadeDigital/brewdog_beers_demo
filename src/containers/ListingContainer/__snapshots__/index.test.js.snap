// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListingContainer /> snapshot: renders default state 1`] = `
<DocumentFragment>
  <div>
    <a
      href="/"
    >
      <header>
        <img
          alt="Brewdog logo"
          role="presentation"
          src="brewdog_logo.png"
        />
        <h1
          class=""
          data-testid="heading"
        >
          Brewdog Beer Collection
        </h1>
      </header>
    </a>
    <div>
      <label
        for="filterDropdown"
      >
        Order by:
      </label>
      <select
        data-testid="dropdown"
        id="filterDropdown"
      >
        <option>
          - Select an option -
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="first_brewed"
        >
          Date first brewed
        </option>
        <option
          value="abv"
        >
          ABV
        </option>
        <option
          value="ibu"
        >
          IBU
        </option>
      </select>
    </div>
    <div>
      <article
        data-testid="card"
      >
        <a
          href="/detail/1"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Buzz
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              A Real Bitter Experience.
            </h3>
          </header>
          <p>
            First brewed: 09/2007
          </p>
          <span>
            ABV: 4.5
          </span>
          <span>
            IBU: 60
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/2"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Trashy Blonde
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              You Know You Shouldn't
            </h3>
          </header>
          <p>
            First brewed: 04/2008
          </p>
          <span>
            ABV: 4.1
          </span>
          <span>
            IBU: 41.5
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/3"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Berliner Weisse With Yuzu - B-Sides
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              Japanese Citrus Berliner Weisse.
            </h3>
          </header>
          <p>
            First brewed: 11/2015
          </p>
          <span>
            ABV: 4.2
          </span>
          <span>
            IBU: 8
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/4"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Pilsen Lager
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              Unleash the Yeast Series.
            </h3>
          </header>
          <p>
            First brewed: 09/2013
          </p>
          <span>
            ABV: 6.3
          </span>
          <span>
            IBU: 55
          </span>
        </a>
      </article>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ListingContainer /> snapshot: renders ordered by abv 1`] = `
<DocumentFragment>
  <div>
    <a
      href="/"
    >
      <header>
        <img
          alt="Brewdog logo"
          role="presentation"
          src="brewdog_logo.png"
        />
        <h1
          class=""
          data-testid="heading"
        >
          Brewdog Beer Collection
        </h1>
      </header>
    </a>
    <div>
      <label
        for="filterDropdown"
      >
        Order by:
      </label>
      <select
        data-testid="dropdown"
        id="filterDropdown"
      >
        <option>
          - Select an option -
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="first_brewed"
        >
          Date first brewed
        </option>
        <option
          value="abv"
        >
          ABV
        </option>
        <option
          value="ibu"
        >
          IBU
        </option>
      </select>
    </div>
    <div>
      <article
        data-testid="card"
      >
        <a
          href="/detail/2"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Trashy Blonde
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              You Know You Shouldn't
            </h3>
          </header>
          <p>
            First brewed: 04/2008
          </p>
          <span>
            ABV: 4.1
          </span>
          <span>
            IBU: 41.5
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/3"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Berliner Weisse With Yuzu - B-Sides
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              Japanese Citrus Berliner Weisse.
            </h3>
          </header>
          <p>
            First brewed: 11/2015
          </p>
          <span>
            ABV: 4.2
          </span>
          <span>
            IBU: 8
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/1"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Buzz
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              A Real Bitter Experience.
            </h3>
          </header>
          <p>
            First brewed: 09/2007
          </p>
          <span>
            ABV: 4.5
          </span>
          <span>
            IBU: 60
          </span>
        </a>
      </article>
      <article
        data-testid="card"
      >
        <a
          href="/detail/4"
        >
          <header
            data-testid="cardHeading"
          >
            <h2
              class=""
              data-testid="heading"
            >
              Pilsen Lager
            </h2>
            <h3
              class=""
              data-testid="heading"
            >
              Unleash the Yeast Series.
            </h3>
          </header>
          <p>
            First brewed: 09/2013
          </p>
          <span>
            ABV: 6.3
          </span>
          <span>
            IBU: 55
          </span>
        </a>
      </article>
    </div>
  </div>
</DocumentFragment>
`;
